Add support for FP16 to openBLAS and shgemm on RISCV

Srangrang opened the pull request:
-add HFLOAT16 and BUILD_HFLOAT16 macro define to distinguish BFLOAT16 and BUILD_BFLOAT16
-add shgemm for RISCV_ZVL128B and RISCV_ZVL256B
-using fp16 on RISCV requires zfh and zvfh instruction sets
-enable fp16 support in Makefile.rule

Related to issue #5279 
Co-authored-by Ao Dong

Srangrang committed:
add shgemm for RISCV_ZVL128B

Srangrang committed:
Add FP16 support for RISCV

Commit:
Added shgemm_kernel_8x8 for RISCV64_ZVL128B and shgemm_kernel_16x8 for RISCV64_ZVL256B

Added HFLOAT16 support for RISCV64
Added shgemm_kernel_8x8 for RISCV64_ZVL128B and shgemm_kernel_16x8 for RISCV64_ZVL256B based on HFLOAT16
The instruction sets used are ZVFH and ZFH, which need to be supported by RVV1.0

Related to issue #5279
Co-authored-by Linjin Li <linjin_li@163.com>

Srangrang committed:
Merge pull request #1 from gkdddd/riscv_shgemm

Added shgemm_kernel_8x8 for RISCV64_ZVL128B and shgemm_kernel_16x8 fo…

Srangrang committed:
fix: resolve the compilation failure without zfh instruction

- modify the macro conditions in Makefile.system
- Delete development test code

Related to issue#5279

Srangrang committed:
Merge branch 'develop' of https://github.com/Srangrang/OpenBLAS into develop

martin-frbg commented:
I _think_ we'll need some adjustments in interface/gemm.c to disable the "small matrix" pathway, adjust the function name used in error messages, and also keep hfloat16 out of the optimum thread number computations, now that it is separate from bfloat16. Please see attached (and add this or similar to the PR if you agree)
[gemm.c.txt](https://github.com/user-attachments/files/20716611/gemm.c.txt)

Also it probably makes sense to seperate the two types to such an extent that one can be built without the other - I have further modified your version of gensymbol(.pl) to take another parameter for the BUILD_HFLOAT16, and adjusted exports/Makefile accordingly:
[Makefile.txt](https://github.com/user-attachments/files/20716654/Makefile.txt)
[gensymbol.pl.txt](https://github.com/user-attachments/files/20716656/gensymbol.pl.txt)
[gensymbol.txt](https://github.com/user-attachments/files/20716661/gensymbol.txt)

martin-frbg commented:
In similar vein, the affected files in the benchmark folder probably need adjusting as well, please check
[Makefile.txt](https://github.com/user-attachments/files/20716702/Makefile.txt)

[gemm.c.txt](https://github.com/user-attachments/files/20716704/gemm.c.txt)

martin-frbg commented:
Lastly, I found that I cannot build the "generic" replacements on a non-RISCV host unless I remove the `ifneq ($(SHGEMM_UNROLL_M), $(SHGEMM_UNROLL_N))` conditional around the compiler commands for the SHGEMMINCOPYOBJ and SHGEMMITCOPYOBJ in kernel/Makefile.L3 (and I wonder if it would make sense to provide hfloat16 conversion functions in kernel/generic/gemm_kernel2x2.c similar to what is present there for bfloat16) ?

Unfortunately, I have not had the time to try a CMake build yet.

martin-frbg commented:
cmake/utils.cmake
[utils.cmake.txt](https://github.com/user-attachments/files/20741995/utils.cmake.txt)
cmake/kernel.cmake
[kernel.cmake.txt](https://github.com/user-attachments/files/20741997/kernel.cmake.txt)
kernel/CMakeLists 
[CMakeLists.txt](https://github.com/user-attachments/files/20742181/CMakeLists.txt)
interface/CMakeLists
[CMakeLists.txt](https://github.com/user-attachments/files/20742183/CMakeLists.txt)

toplevel CMakeLists (added status messages, updated gensymbol calls, updated benchmark build to include B/HFLOAT16)
[CMakeLists.txt](https://github.com/user-attachments/files/20742002/CMakeLists.txt)

Srangrang commented:
> I _think_ we'll need some adjustments in interface/gemm.c to disable the "small matrix" pathway, adjust the function name used in error messages, and also keep hfloat16 out of the optimum thread number computations, now that it is separate from bfloat16. Please see attached (and add this or similar to the PR if you agree) [gemm.c.txt](https://github.com/user-attachments/files/20716611/gemm.c.txt)
> 
> Also it probably makes sense to seperate the two types to such an extent that one can be built without the other - I have further modified your version of gensymbol(.pl) to take another parameter for the BUILD_HFLOAT16, and adjusted exports/Makefile accordingly: [Makefile.txt](https://github.com/user-attachments/files/20716654/Makefile.txt) [gensymbol.pl.txt](https://github.com/user-attachments/files/20716656/gensymbol.pl.txt) [gensymbol.txt](https://github.com/user-attachments/files/20716661/gensymbol.txt)

Thank you for your comment, @martin-frbg . I agree with your modification and will incorporate your changes into this PR.

Srangrang commented:
> In similar vein, the affected files in the benchmark folder probably need adjusting as well, please check [Makefile.txt](https://github.com/user-attachments/files/20716702/Makefile.txt)
> 
> [gemm.c.txt](https://github.com/user-attachments/files/20716704/gemm.c.txt)

I have reviewed the contents of the benchmark folder and noticed that references to BF16 appear to be defined as `HALF`. However, the modifications provided in your Makefile.txt suggest changing this to `BFLOAT16`. I believe this change could reduce ambiguity and more effectively differentiate between FP16 and BF16. Could we consider standardizing the use of `HFLOAT16` for FP16 and `BFLOAT16` for BF16 within OpenBLAS (by modifying the definition of `GOTO_HF16_TARGETS` to `GOTO_HFLOAT16_TARGETS` in this file)?

Additionally, I have identified other instances where `HALF` is used to define BF16, such as in the driver/level3/Makefile. Should we also adjust these to use the `BFLOAT16` definition?

martin-frbg commented:
You are right - any occurences of HALF must be leftovers from the initial BFLOAT16 code contribution that was using "half precision" and the "h" prefix throughout.

Srangrang committed:
fix: resolve non-RISCV host build failed issue

- adjust interface to disable "small matrix" pathway
- separate HFLOAT16 from BFLOAT16
- remove SHGEMM_UNROLL_M and SHGEMM_UNROLL_N equal conditions

Related to PR#5290
Co-authored-by Martin

Srangrang committed:
style: modify HALF to BFLOAT16 in benchmark folder

Srangrang committed:
disable BUILD_HFLOAT16

ChipKerchner commented:
I believe the vector instructions within the if ( M & 4 ) in 128b can be m1 versus what they are currently m2.

ChipKerchner commented on kernel/riscv64/shgemm_kernel_8x8_zvl128b.c:
```diff
@@ -0,0 +1,767 @@
+            vfloat16m1_t A0 = __riscv_vle16_v_f16m1(&A[ai + 0 * gvl], gvl);
+            ai += 4;
+
+            vfloat32m2_t result0 = __riscv_vfwmul_vf_f32m2( A0, B0, gvl);
```
I believe the vector instructions within the if ( M & 4 ) in 128b can be m1 versus what they are currently m2.

Srangrang replied:
Thanks again for the detailed review — and especially for resolving this issue in #5451!

ChipKerchner commented on kernel/riscv64/shgemm_kernel_16x8_zvl256b.c:
```diff
@@ -0,0 +1,969 @@
+            BLASLONG bi = n_top * K;
+            
+            for(BLASLONG k=0; k<K; k++) {
+                result0+=(float)(A[ai+0]*B[bi+0]);
```
I believe this is wrong. A * B will produce a _Float16 result before it is cast. _Float16 is a storage type (at least before C++23) - unsigned short. So A * B is unsigned short * unsigned short and then cast to float.

It should be

result0+=(float)(A[ai+0])*(float)(B[bi+0]);

Srangrang replied:
Thanks for pointing this out!

ChipKerchner commented on kernel/riscv64/KERNEL.RISCV64_ZVL128B:
```diff
@@ -245,3 +245,14 @@ endif
 ZGEMM_BETA = zgemm_beta_rvv.c
 endif
+
+ifeq ($(BUILD_BFLOAT16), 1)
```
Bug here - should be BUILD_HFLOAT16.  Fixed in MR 5454.

ChipKerchner commented on kernel/riscv64/KERNEL.RISCV64_ZVL256B:
```diff
@@ -209,5 +209,23 @@ COMATCOPY_CN = zomatcopy_cn_vector.c
 SOMATCOPY_CN = omatcopy_cn_vector.c
 
+
+ifeq ($(BUILD_BFLOAT16), 1)
```
Bug here - should be BUILD_HFLOAT16.  Fixed in MR 5454.

Srangrang replied:
Indeed, thanks.
